crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -1180,6 +1181,252 @@ impl<'w> EntityWorldMut<'w> {
+    ///
+    /// // Initial component tree:
+    /// //  X
+    /// //   \
+    /// //    Y   W
+    /// //     \ /
+    /// //      Z
+    ///
+    /// assert!(world.entity(entity).contains::<X>());
+    /// assert!(world.entity(entity).contains::<Y>());
+    /// assert!(world.entity(entity).contains::<Z>());
+    /// assert!(world.entity(entity).contains::<W>());
+    ///
+    /// // Remove X and its unused requirements
+    /// world.entity_mut(entity).require_recursive_remove::<X>();
+    ///
+    /// // Resulting component tree:
+    /// //     W
+    /// //    /
+    /// //   Z
+    ///
+    /// assert!(!world.entity(entity).contains::<X>());
+    /// assert!(!world.entity(entity).contains::<Y>());
+    /// assert!(world.entity(entity).contains::<Z>());
+    /// assert!(world.entity(entity).contains::<W>());
+    /// ```
+    /// Note: It will note delete components in [`Bundle`] if some components require bundle components. Use
+    ///
+    pub fn require_recursive_remove<T: Bundle>(&mut self) -> &mut Self {
+        let storages = &mut self.world.storages;
+        let components = &mut self.world.components;
+        let bundle = self.world.bundles.init_info::<T>(components, storages);
+
+        // SAFETY: the `BundleInfo` is initialized above
+        // We make clone of components array to not lock immutable access to world
+        let contributed_components = unsafe {
+            self.world
+                .bundles
+                .get_unchecked(bundle)
+                .contributed_components()
+                .to_vec()
+        };
+
+        let old_location = self.location;
+        let old_archetype = &mut self.world.archetypes[old_location.archetype_id];
+
+        let remaining_components = old_archetype
+            .components()
+            .filter(|c| !contributed_components.contains(c))
+            .collect::<Vec<_>>();
+
+        let remaining_bundle = self
+            .world
+            .bundles
+            .init_dynamic_info(components, &remaining_components);
+
+        // SAFETY: the `BundleInfo` is initialized above
+        let remaining_bundle_info = unsafe { self.world.bundles.get_unchecked(remaining_bundle) };
+
+        // Remove components required by remaining components from contributed_components of bundle T
+        let to_delete = contributed_components
+            .iter()
+            .filter(|c| !remaining_bundle_info.contributed_components().contains(c))
+            .copied()
+            .collect::<Vec<_>>();
+
+        let to_delete_bundle = self.world.bundles.init_dynamic_info(components, &to_delete);
+        // SAFETY: the dynamic `BundleInfo` is initialized above
+        self.location = unsafe { self.remove_bundle(to_delete_bundle) };
+
+        self
+    }
+
+    /// Removes the components of this [`Bundle`], components that require them, and their required components,
+    /// unless they are required by components outside of this removal set.
```

Comment:
It took me a while (and the example) to understand what this function does, and even then I'm not sure what happens in all cases. It also seems to me this is going a bit out of scope. Is there an actual usecase for this?

Reply:
> It took me a while (and the example) to understand what this function does, and even then I'm not sure what happens in all cases. It also seems to me this is going a bit out of scope. Is there an actual usecase for this?

I can think of two examples of usage:
1. Complex status effects in games: When a character is affected by a status (e.g., poisoning), it may involve multiple related components. Different types of poisoning could require different visual effects, all depending on a core "Poisoned" component. When curing the poison, you want to remove not just the core component, but all related effect components. This function allows you to remove the entire "tree" of components related to the poison effect without having to manually track and remove each one.
2. Library cleanup for modular systems: In a scenario where you're developing a library (e.g., for animations), end users might extend your core components with their own custom components. When an animation completes and you want to clean up, using require_descendant_remove on the core animation component would also remove all user-defined components that depend on it, ensuring a thorough cleanup without leaving orphaned components behind.